sparse/file.go

```diff
@@ -192,46 +194,186 @@ func HashData(data []byte) ([]byte, error) {
 	return sum[:], nil
 }
 
-func GetFiemapExtents(file FileIoProcessor) ([]Extent, error) {
+func GetFileLayout(ctx context.Context, file FileIoProcessor) (<-chan FileInterval, <-chan error, error) {
 	fileInfo, err := file.Stat()
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
-	return GetFiemapRegionExts(file, Interval{0, fileInfo.Size()})
+
+	const MaxExtentsBuffer = 1024
+	const MaxInflightIntervals = MaxExtentsBuffer * 10
+	out := make(chan FileInterval, MaxInflightIntervals)
+	errc := make(chan error, 1)
+	go func() {
+		defer close(out)
+		defer close(errc)
```

Comment:
don't understand what you mean?

Reply:
For example, if getting fiemap extents failed and the goroutine returns, is it possible that the error channel may get closed before the receiver handle the error signal (the receiver will check if the channel is closed first)? If yes, then the error will be lost and the receivers wrongly consider everything works fine.
Similar concern for the last signal of the interval channel.

Reply:
This should not be a concern, let me try walking you through your mentioned case, on the example of the errc channel, same applies to the out channel.

```go
	out, errc, err := GetFileLayout(ctx, file)
        // err itself is prestart of the go routine only, so nothing to worry about, since we will exit

	// let's say the go routine has nothing todo (fileInfo.Size() == 0) case
	// so it returns immediately, and closes the errc, out channels that we assigned above
	// I am guessing this is your concerned case correct?

	const WorkerCount = 4
	errorChannels := []<-chan error{errc}
	for i := 0; i < WorkerCount; i++ {
		errorChannels = append(errorChannels, processFileIntervals(ctx, out, processSegment))
	}

	// since out is closed already as well, the individual processFileInterval go routines have nothing to do and exit as well
	// you can think through both cases, they will close later or are closed already here, this makes no difference in regard
	// to the correctness of the solution, so let's go with the they have closed their error channels already

	// the below select will exit once all error channels are closed, or a single
	// channel has run into an error, which will lead to the ctx being cancelled
	mergedErrc := mergeErrorChannels(ctx, errorChannels...)
	select {
	case err = <-mergedErrc:
		break
	}

	// above we have merged all the error channels and are waiting for a signal on the mergedErrc out channel.
        // let's have a look at how the mergeErrorChannels actuall work
        return err
```
-------------
```go
// mergeErrorChannels will merge all error channels into a single error out channel.
// the error out channel will be closed once the ctx is done or all error channels are closed
// if there is an error on one of the incoming channels the error will be relayed.
func mergeErrorChannels(ctx context.Context, channels ...<-chan error) <-chan error {
	var wg sync.WaitGroup
	wg.Add(len(channels))

	// here we create a buffered output channel, for each go routine one slot
	// for the worst case scenario where each go routines errors at the same time
	// this also prevents them from getting stuck trying to send an error
	out := make(chan error, len(channels))
	output := func(c <-chan error) {
		defer wg.Done()
		select {
		case err, ok := <-c:
			if ok {
				out <- err
			}
			return
		case <-ctx.Done():
			return
		}
	}

	// create a relaying go routine for each of the incoming error channels
	// that rely the error into the output channel and exit
	for _, c := range channels {
		go output(c)
	}

	// we wait till all go routines have exited, either because they had an error
	// the ctx was cancelled as a result of a single error
	// or all the incoming error channels where closed (this is the case that you are concerned about)
	// afterwards we close our error channel and exit this go routine.
	go func() {
		wg.Wait()
		close(out)
	}()
	return out
}
```

------------------

As you see there is should no be any issue.
I think your concern stems from the fact that you are concerned about reading from a closed channel:
- one will continuously read the default value from a closed channel, so it's fine reading from a closed channel
- one can use the two return form to determine whether a channel is closed `case err, open := <-c:`
- all producer go routines use a buffered errc[1] channel, so they don't block
- buffered channels are drained first in / first out (d1,d2, close, close, close, close, close, etc)

On any error the ctx will be cancelled so all go routines will terminate and each channel closed.
Once all go routines are closed, there will be no more references to the channels and they can be garbage collected even if they still have unprocessed data (early exit interval processing do to error)

-----------------

Let me know if that helps clarify the situation, if not we can do a peer review and discuss it while going over the code.

Reply:
Got it. Reading a closed channel should not be an issue. I am sorry for misremembering this golang mechanism. Your explanation is pretty clear.

Reply:
Not a problem, good that you asked. 
What is an issue is if you are reading from a nil channel since that will always block.
As well as sending to a closed/nil channel which will panic.